nexus/src/app/background/tasks/tuf_artifact_replication.rs

```
@@ -0,0 +1,447 @@
+        depot_sled: &'a Sled,
+        hash: ArtifactHash,
+    },
+    Delete {
+        sled: &'a Sled,
+        hash: String,
+    },
+}
+
+impl BackgroundTask for ArtifactReplication {
+    fn activate<'a>(
+        &'a mut self,
+        opctx: &'a OpContext,
+    ) -> BoxFuture<'a, serde_json::Value> {
+        async {
+            match self.activate_internal(opctx).await {
+                Ok(status) => serde_json::to_value(status).unwrap(),
+                Err(err) => {
+                    let err_string = format!("{:#}", err);
+                    error!(
+                        &opctx.log,
+                        "error during artifact replication";
+                        "error" => &err_string,
+                    );
+                    serde_json::json!({
+                        "error": err_string,
+                    })
+                }
+            }
+        }
+        .boxed()
+    }
+}
+
+impl ArtifactReplication {
+    pub fn new(
+        datastore: Arc<DataStore>,
+        local_rx: mpsc::Receiver<ArtifactsWithPlan>,
+    ) -> ArtifactReplication {
+        ArtifactReplication { datastore, local: Vec::new(), local_rx }
+    }
+
+    async fn activate_internal<'a>(
+        &'a mut self,
+        opctx: &'a OpContext,
+    ) -> anyhow::Result<TufArtifactReplicationStatus> {
+        let log = &opctx.log;
+        let datastore = &self.datastore;
+
+        // Move any received artifacts out of `local_rx` into `local`.
+        loop {
+            match self.local_rx.try_recv() {
+                Ok(artifacts) => self.local.push(artifacts),
+                Err(TryRecvError::Empty) => break,
+                Err(TryRecvError::Disconnected) => {
+                    // If the last sender for this channel is dropped, then the
+                    // `Nexus` type has also been dropped. This is presumably
+                    // a bug.
+                    panic!("artifact replication receiver disconnected");
```

Nit: I know it's unlikely, but can we return an error here instead? This behavior makes implementing clean-shutdown really painful